### ISSUE TITLE
502 is happen when "Data Sources / Elasticsearch", "URL: https://localhost:9200/", "Basic Auth: on" and "Skip TLS Verify: on" (Version: 6.5.0 and 6.5.1)

### ISSUE THREAD

**Issue by t-umeno** [grafana/grafana#20944]
**What happened**:
502 is happen when "Data Sources / Elasticsearch", "URL: https://localhost:9200/", "Basic Auth: on" and "Skip TLS Verify: on" (Version: 6.5.0 and 6.5.1)
**What you expected to happen**:
Index OK. Time field name OK. (Version: 6.4.5)
**How to reproduce it (as minimally and precisely as possible)**:
Use opendistro for Elasticsearch  https://opendistro.github.io/for-elasticsearch/
**Anything else we need to know?**:
none

**Environment**:
- Grafana version: 6.5.0 and 6.5.1
- Data source type & version: opendistro for Elasticsearch & 1.2.1 or any version
- OS Grafana is installed on: Ubuntu 18.04 LTS
- User OS & Browser: macOS 10.13.6 & firefox 71.0
- Grafana plugins: none
- Others: none

**Comment by zhiguangwang** [comment]
Exact same issue here.

Had to rollback to v6.4.4

**Comment by aknuds1** [comment]
Can you tell us the command with which to run a Docker container of amazon/opendistro-for-elasticsearch, in order to reproduce, please? Also, are there any default basic auth creds we can use?

**Comment by aknuds1** [comment]
I tried running Elasticsearch in Docker as follows: `docker run -p 9200:9200 -p 9600:9600 -e "discovery.type=single-node"  amazon/opendistro-for-elasticsearch:1.2.1`, then adding it as a data source in Grafana v6.4.5 (https://localhost:9200, basic auth enabled w/ creds admin/admin, and skip TLS verify enabled). It fails with 502 error, apparently due to a TLS handshake error. What's your setup to make Grafana v6.4 connect successfully to Elastic?

**Comment by t-umeno** [comment]
I use default basic auth creds we can use (admin/admin).

**Comment by aknuds1** [comment]
Thanks @t-umeno - can you tell me though which command you use to run Elasticsearch? As I said, I tried the following and can not add it as a data source to Grafana v6.4.5: `docker run -p 9200:9200 -p 9600:9600 -e "discovery.type=single-node" amazon/opendistro-for-elasticsearch:1.2.1`.

**Comment by t-umeno** [comment]
Please see https://github.com/t-umeno/visualize_network_traffic

# visualize_network_traffic
* visualize network traffic flow use IPFIX form YAF by opendistro for Elasticsearch
* install opendistro for elasticsearch, YAF, super_mediator, curl and wireshark on ubuntu/bionic64
## requirements
* more than 8GB memory (better over 12GB memory) Mac, Windows, Linux
* VirtualBox
* Vagrant
* vagrant plugin install vagrant-disksize
* vagrant plugin install vagrant-proxyconf (if you use proxy)
* vagrant plugin install vagrant-vbguest (option)

### install
Please set http_proxy and https_proxy if you need.
```
$ git checkout https://github.com/t-umeno/visualize_network_traffic.git
$ cd visualize_network_traffic
$ echo 'elasticstack_system: opendistro_docker' > ansible/ELK6/playbooks/settings/config.yml
$ cd ansible/ELK6/playbooks
$ vagrant up
```

**Comment by t-umeno** [comment]
Select interface for packet capture if display prompt

```
(snip)
(select network interface for packet capture)
==> default: Available bridged network interfaces:
1) en1: Wi-Fi (AirPort)
2) bridge0
3) en0: Ethernet
4) en2: Thunderbolt 1
5) p2p0
==> default: When choosing an interface, it is usually the one that is
==> default: being used to connect to the internet.
default: Which interface should the network bridge to? 3
(snip)
```

Please check json file in ~/yaf/json

```
$ vagrant ssh
(snip)
vagrant@ubuntu-xenial:~$ ls -l yaf/json
total 2396
-rw-r--r-- 1 root root 2447363 Dec 18 06:11 yaf.20181217154009.json
vagrant@ubuntu-xenial:~$
```

This method will install grafana 6.4.5

**Comment by t-umeno** [comment]
This method will install opendistro for Elasticsearch 1.2.1 (docker) too.

## usage (grafana)
- open http://127.0.0.1:3000/
- login (username: admin password: admin)
- change password
- push "Add data source"
- push "Elasticsearch" from "Choose data source type"
- input "yaf" in Name
- input "https://localhost:9200" in URL
- check "Basic Auth: on" 
- check "Skip TLS Verify: on" 
- input "yaf.*" in Elasticsearch details Index Name
- select "7.0+" in Elasticsearch deaiils Version
- push "Save & Test"

 Index OK. Time field name OK. (grafana 6.4.5)

**Comment by t-umeno** [comment]
If you edit in ansible/ELK6/playbooks/settings/version.yml

from:
```
grafana_version_with_opendistro: 6.4.5
```

to:
```
grafana_version_with_opendistro: 6.5.0
```

```
$ vagrant destroy
$ vagrant up
```

502 will be happened by same operation

https://github.com/t-umeno/visualize_network_traffic/blob/master/ansible/ELK6/playbooks/settings/version.yml

**Comment by t-umeno** [comment]
Please check template of /usr/local/etc/opendistro/docker-compose.yml

https://github.com/t-umeno/visualize_network_traffic/blob/master/ansible/ELK6/playbooks/roles/opendistro/templates/docker-compose_elasticsearch_1node.yml.j2

Thanks!

**Comment by t-umeno** [comment]
Please read URL:

https://opendistro.github.io/for-elasticsearch-docs/
https://opendistro.github.io/for-elasticsearch-docs/docs/install/docker/

```
docker pull amazon/opendistro-for-elasticsearch:1.2.1
docker run -p 9200:9200 -p 9600:9600 -e "discovery.type=single-node" amazon/opendistro-for-elasticsearch:1.2.1
```

**Comment by t-umeno** [comment]
502 is happened (Version 6.5.2)
But, another environment (http_proxy and https_proxy are set)
Index OK. Time field name OK (Version 6.5.2), why?

**Comment by marefr** [comment]
@t-umeno hard to tell. Feels like non-Grafana related to me since it works in one environment. Are you using Server or Browser access mode in data source config in this and other environment? 

Maybe related to that you're using a proxy in that other environment - Suppose other environment are not using localhost as url to elasticsearch? 

Can you use chrome developer tools or similar and check network activity when you receive the 502 and include full request/response plus headers here?

**Comment by t-umeno** [comment]
I use Server access mode in data source config in this and other environment.

**Comment by t-umeno** [comment]
> Suppose other environment are not using localhost as url to elasticsearch?
No, I use https://localhost:9200/ in this environment, too

**Comment by t-umeno** [comment]
@marefr
I'm sorry to have kept you waiting.

Grafana 6.5.3 and Chrome 79.0.3945.88 (macOS 10.13.6)
Chrome devtools log:
```
DashboardPage.d9f582b9b538bf193c78.js:1 [Violation] 'setTimeout' handler took 55ms
angular~app.d9f582b9b538bf193c78.js:7 [Violation] 'setTimeout' handler took 50ms
angular.js:12845 GET http://127.0.0.1:3000/api/datasources/proxy/1/yaf.*/_mapping 502 (Bad Gateway)
(anonymous) @ angular.js:12845
(anonymous) @ angular.js:12572
(anonymous) @ angular.js:12320
(anonymous) @ angular.js:17169
(anonymous) @ angular.js:17217
$digest @ angular.js:18351
(anonymous) @ angular.js:18581
p @ angular.js:6428
(anonymous) @ angular.js:6707
setTimeout (async)
i.defer @ angular.js:6705
$evalAsync @ angular.js:18579
(anonymous) @ angular.js:17041
f @ angular.js:17217
then @ angular.js:17138
v @ angular.js:12236
e.datasourceRequest @ backend_srv.ts:236
t.request @ datasource.ts:83
t.get @ datasource.ts:95
t.getFields @ datasource.ts:383
t.testDatasource @ datasource.ts:279
(anonymous) @ DataSourceSettingsPage.tsx:155
(anonymous) @ tslib.es6.js:99
(anonymous) @ tslib.es6.js:80
(anonymous) @ tslib.es6.js:73
l @ tslib.es6.js:69
(anonymous) @ DataSourceSettingsPage.tsx:153
e.withNoBackendCache @ backend_srv.ts:103
(anonymous) @ DataSourceSettingsPage.tsx:153
(anonymous) @ tslib.es6.js:99
(anonymous) @ tslib.es6.js:80
a @ tslib.es6.js:70
Promise.then (async)
s @ tslib.es6.js:72
(anonymous) @ tslib.es6.js:73
l @ tslib.es6.js:69
t.testDataSource @ DataSourceSettingsPage.tsx:144
(anonymous) @ DataSourceSettingsPage.tsx:102
(anonymous) @ tslib.es6.js:99
(anonymous) @ tslib.es6.js:80
a @ tslib.es6.js:70
Promise.then (async)
s @ tslib.es6.js:72
(anonymous) @ tslib.es6.js:73
l @ tslib.es6.js:69
a.onSubmit @ DataSourceSettingsPage.tsx:97
onSubmit @ DataSourceSettingsPage.tsx:292
onClick @ ButtonRow.tsx:19
(anonymous) @ react-dom.production.min.js:14
h @ react-dom.production.min.js:15
(anonymous) @ react-dom.production.min.js:15
k @ react-dom.production.min.js:17
T @ react-dom.production.min.js:18
E @ react-dom.production.min.js:18
P @ react-dom.production.min.js:21
Sn @ react-dom.production.min.js:84
Iu @ react-dom.production.min.js:250
Rt @ react-dom.production.min.js:39
Tn @ react-dom.production.min.js:86
(anonymous) @ react-dom.production.min.js:251
e.unstable_runWithPriority @ scheduler.production.min.js:18
Nu @ react-dom.production.min.js:251
Cn @ react-dom.production.min.js:85
datasource.ts:290 {data: "", status: 502, config: {…}, statusText: "Bad Gateway", headers: ƒ, …}data: ""status: 502headers: ƒ (n)arguments: (...)caller: (...)length: 1name: ""prototype: {constructor: ƒ}__proto__: ƒ ()[[FunctionLocation]]: angular.js:11452[[Scopes]]: Scopes[3]config: method: "GET"transformRequest: [ƒ]transformResponse: [ƒ]paramSerializer: ƒ (t)jsonpCallbackParam: "callback"url: "api/datasources/proxy/1/yaf.*/_mapping"data: undefinedretry: 0headers: {X-Grafana-Org-Id: 1, X-Grafana-NoCache: "true", Accept: "application/json, text/plain, */*"}__proto__: ObjectstatusText: "Bad Gateway"xhrStatus: "complete"__proto__: constructor: ƒ Object()__defineGetter__: ƒ __defineGetter__()__defineSetter__: ƒ __defineSetter__()hasOwnProperty: ƒ hasOwnProperty()__lookupGetter__: ƒ __lookupGetter__()__lookupSetter__: ƒ __lookupSetter__()isPrototypeOf: ƒ isPrototypeOf()propertyIsEnumerable: ƒ propertyIsEnumerable()toString: ƒ toString()valueOf: ƒ valueOf()toLocaleString: ƒ toLocaleString()get __proto__: ƒ __proto__()set __proto__: ƒ __proto__()
```

**Comment by marefr** [comment]
@t-umeno 502 bad gateway is an indication that Grafana server can't establish a connection with your elasticsearch cluster. Seems like configuration error to me and not basic auth related because of 502 status, i.e. Grafana can't connect at all to Elasticsearch.

**Comment by t-umeno** [comment]
Index OK. Time field name OK. (Version: 6.5.3)
Thanks!

**Comment by zhiguangwang** [comment]
The root causes are:

- Grafana >= 6.5.0 is [built with Go 1.13](https://community.grafana.com/t/release-notes-v6-5-x/22704), [which enables TLS 1.3 by default](https://golang.org/doc/go1.13#tls_1_3)
- [A bug in TLS 1.3 implementation](https://bugs.openjdk.java.net/browse/JDK-8210334) of Java 11.0.1 (fixed in 11.0.2)
- Unfortunately, [latest ODFE still uses Java 11.0.1](https://github.com/opendistro-for-elasticsearch/opendistro-build/blob/master/elasticsearch/docker/templates/Dockerfile.j2#L40)

Basically, you have 3 workarounds:

1. Disable TLS 1.3 on Grafana >= 6.5.0 by adding the value `tls13=0` to the `GODEBUG` environment variable
2. Disable TLS 1.3 on ODFE with `opendistro_security.ssl.http.enabled_protocols` in [elasticsearch.yml](https://opendistro.github.io/for-elasticsearch-docs/docs/troubleshoot/tls/#tls-versions)
3. Bump Java to at least 11.0.2 when running ODFE on Docker (mount another JDK at `/opt/jdk-11.0.1`)

**Comment by marefr** [comment]
@zhiguangwang thank you for the investigation!

**Comment by derwin12** [comment]
I came across this bug note .. I believe Grafana 7.x is not built with GO 1.14 which removed this GODEBUG workaround .. or atleast it seem to not work for me. I am rolling back to grafana 6.x to see if I can get it working.
